src/finn/transformation/fpgadataflow/templates.py

```diff
@@ -126,120 +126,113 @@ def __init__(
         N = self.N
-        ishape_normal = $INPUT_SHAPE_NORMAL$
         # load desired input .npy file
-        ibuf_normal = np.load(self.inputfile)
+        ibuf_normal = np.load(inputfile)
         # ensure that shape is as expected
-        assert ibuf_normal.shape == ishape_normal
+        assert ibuf_normal.shape == self.ishape_normal
         return ibuf_normal
 
     def pack_input(self, ibuf_normal):
         N = self.N
-        # input FINN DataType
-        idt = $INPUT_FINN_DATATYPE$
-        ishape_folded = $INPUT_SHAPE_FOLDED$
         # convert to folded form
-        ibuf_folded = ibuf_normal.reshape(ishape_folded)
+        ibuf_folded = ibuf_normal.reshape(self.ishape_folded)
         # pack the input buffer, reversing both SIMD dim and endianness
         ibuf_packed = finnpy_to_packed_bytearray(
-            ibuf_folded, idt, reverse_endian=True, reverse_inner=True
+            ibuf_folded, self.idt, reverse_endian=True, reverse_inner=True
         )
         return ibuf_packed
 
     def unpack_output(self, obuf_packed):
         N = self.N
-        # output FINN DataType
-        odt = $OUTPUT_FINN_DATATYPE$
-        oshape_folded = $OUTPUT_SHAPE_FOLDED$
         # unpack the packed output buffer from accelerator
         obuf_folded = packed_bytearray_to_finnpy(
-            obuf_packed, odt, oshape_folded, reverse_endian=True, reverse_inner=True
+            obuf_packed, self.odt, self.oshape_folded, reverse_endian=True, reverse_inner=True
         )
         return obuf_folded
 
-    def save_output(self, obuf_folded):
+    def save_output(self, obuf_folded, outputfile):
         N = self.N
         # convert to normal reshape and save
-        oshape_normal = $OUTPUT_SHAPE_NORMAL$
-        obuf_normal = obuf_folded.reshape(oshape_normal)
-        np.save(self.outputfile, obuf_normal)
-
-    def allocate_pynqbuffer(self, shape, data=None):
-        buf_device = allocate(shape=shape, dtype=np.uint8)
+        obuf_normal = obuf_folded.reshape(self.oshape_normal)
+        np.save(outputfile, obuf_normal)
 
+    def allocate_pynqbuffers(self, data=None):
+        ibuf_packed_device = allocate(shape=self.ishape_packed, dtype=np.uint8)
         # if necessary copy the packed data into the PYNQ buffer
         # TODO optimization: pack directly into the PYNQ buffer?
         if data is not None:
-            np.copyto(buf_device, data)
-
-        return buf_device
+            np.copyto(ibuf_packed_device, data)
+        obuf_packed_device = allocate(shape=self.oshape_packed, dtype=np.uint8)
 
+        return [ibuf_packed_device, obuf_packed_device]
 
-    def run_nw(self):
-        exec_mode = self.exec_mode
-        if exec_mode == "remote_pynq":
-            ibuf_normal = self.load_input()
-            ibuf_packed = self.pack_input(ibuf_normal)
-        elif exec_mode != "throughput_test":
-            raise Exception("Exec mode has to be set to remote_pynq or throughput_test")
-
+    def setup_TLastMarker(self):
```

Comment:
remove this member function or make "private" (usually indicated in Python by an underscore before function name)
our preferred strategy is to allocate buffers as part of driver constructor
changing N after construction implies deallocating and reallocating those buffers, let's not do that for now (let N remain constant once instance is constructed)
do all tlastmarker setup inside constructor